beater/beater.go

```diff
@@ -186,6 +188,12 @@ func initTracer(info beat.Info, config *Config, logger *logp.Logger) (*elasticap
 	} else {
 		os.Setenv("ELASTIC_APM_ACTIVE", "true")
 		logger.Infof("self instrumentation is enabled")
+
+		if config.SelfInstrumentation.Hosts != nil {
+			os.Setenv("ELASTIC_APM_SECRET_TOKEN", config.SelfInstrumentation.SecretToken)
+			os.Setenv("ELASTIC_APM_SERVER_URL", config.SelfInstrumentation.Hosts[0])
```

Comment:
These env vars only take effect in a call to `transport.NewHTTPTransport`. `elasticapm.NewTracer` uses `transport.Default`, which is initialised at package-init time.

You would either have to call `transport.InitDefault` after setting these, or alternatively (my preference) call `transport.NewHTTPTransport` with the arguments directly.